alf/algorithms/tasac_algorithm.py

```diff
@@ -0,0 +1,676 @@
+        """Compute Q(s,a)"""
+        observation = (observation, action)
+        critics, _ = critic_net(observation)  # [B, replicas * reward_dim]
+        critics = torch.reshape(  # [B, replicas, reward_dim]
+            critics, (-1, self._num_critic_replicas) + self._reward_spec.shape)
+        if replica_min:
+            if self.has_multidim_reward():
+                sign = self.reward_weights.sign()
+                critics = (critics * sign).min(dim=1)[0] * sign
+            else:
+                critics = critics.min(dim=1)[0]
+
+        if apply_reward_weights and self.has_multidim_reward():
+            critics = critics * self.reward_weights
+            critics = critics.sum(dim=-1)
+        return critics
+
+    def _alpha_train_step(self, beta_entropy, action_entropy):
+        alpha_loss = (self._log_alpha[1] *
+                      (action_entropy - self._target_entropy[1]).detach())
+        alpha_loss += (self._log_alpha[0] *
+                       (beta_entropy - self._target_entropy[0]).detach())
+        return alpha_loss
+
+    def _calc_critic_loss(self, experience, train_info: TasacInfo):
+        # We need to put entropy reward in ``experience.reward`` instead of
+        # ``target_critics`` because in the case of multi-step TD learning,
+        # the entropy should also appear in intermediate steps!
+        if self._use_entropy_reward:
+            with torch.no_grad():
+                actor_extra = train_info.actor.extra
+                beta_alpha = self._log_alpha[0].exp()
+                alpha = self._log_alpha[1].exp()
+                entropy_reward = (beta_alpha * actor_extra.beta_entropy +
+                                  alpha * actor_extra.action_entropy)
+                gamma = self._critic_losses[0].gamma
+                experience = experience._replace(
+                    reward=experience.reward + entropy_reward * gamma)
+
+        critic_info = train_info.critic
+        critic_losses = []
+        for i, l in enumerate(self._critic_losses):
+            kwargs = dict(
+                experience=experience,
+                value=critic_info.critics[:, :, i, ...],
+                target_value=critic_info.target_critic)
+            if isinstance(l, TASACTDLoss):
+                kwargs["train_b"] = train_info.action[0]
+
+            critic_losses.append(l(**kwargs).loss)
+
+        critic_loss = math_ops.add_n(critic_losses)
+        return LossInfo(
+            loss=critic_loss,
+            extra=critic_loss / float(self._num_critic_replicas))
+
+    def _trainable_attributes_to_ignore(self):
+        return ['_target_critic_networks']
+
+    def _compute_beta_and_action(self, observation, b0_action, epsilon_greedy,
+                                 mode):
+        # compute resampling action dist
+        action_dist, _ = self._actor_network((observation.detach(), b0_action))
+        # resample a new attempting action
+        if mode == "predict":
```

Comment:
> mode == Mode.predict

This is actually a bug. I'm not sure when this was introduced and if it affected the experiment results or not. Potentially this will make the eval curve of TASAC slightly worse than what it should be.